src/lib.rs

```diff
@@ -496,28 +506,29 @@ pub fn run(args: &Arguments, mut tests: Vec<Trial>) -> Conclusion {
         }
     } else {
         // Run test in thread pool.
-        let pool = match args.test_threads {
-            Some(num_threads) => ThreadPool::new(num_threads),
-            None => ThreadPool::default()
-        };
+        let num_tests = tests.len();
         let (sender, receiver) = mpsc::channel();
 
-        let num_tests = tests.len();
-        for test in tests {
+        let mut tasks: Vec<pool::BoxedTask> = Default::default();
+
+        for test in tests.into_iter() {
             if args.is_ignored(&test) {
                 sender.send((Outcome::Ignored, test.info)).unwrap();
             } else {
                 let sender = sender.clone();
-                pool.execute(move || {
+
+                tasks.push(Box::new(move || {
                     // It's fine to ignore the result of sending. If the
                     // receiver has hung up, everything will wind down soon
                     // anyway.
                     let outcome = run_single(test.runner, test_mode);
                     let _ = sender.send((outcome, test.info));
-                });
+                }));
             }
         }
 
+        pool::scoped_run_tasks(tasks, num_threads);
+
```

Comment:
The function is pretty short and only used in a single place (here), so why not inline it? Similar to what @hanna-kruppe said above, I think in this case we lose more by obscuring what actually happens compared to what we gain by black box abstraction. 

Further, using `thread::scope` is actually problematic when looking at the next loop: that expects to run while the worker threads are doing something, in order to print something immediately after a test is finished. `thread::scope` waits for all threads to be done before returning. So in your code, all finished tests would fill up the channel, and only after all tests are done does the following loop run and print all tests. 
This can be fixed by moving the printing loop below into the `thread::scope` closure, which also requires inlining the function.

Aaand in that case we can also get rid of the first loop, since we already have a vector of "tasks" (the trials) we can iterate over.

Reply:
I think the way you inlined it in d194ff99f600e9adecd1ffb262e764b9ab8e03d2 has a bug: it spawns `num_threads` scoped threads, each of which only takes *one* trial from `iter` and then terminates. So if you have fewer threads than trials, not all tests are executed.

Reply:
Oh no

Reply:
Ooof, that's awkward. Thanks for catching it! Should be fixed in d6ac84fe44f15bddf801a0a59599995b9363068c. Would you mind reviewing that, to avoid me publishing another avoidable bug?

Reply:
Left a comment on that commit.

Reply:
Thanks for reviewing. I removed the outdated comment in e17efb916a68573063449d40ea368a463703e991 and also added additional tests in 41a18561074fab4602505b5b646fd352ea04ac66 to catch this kind of bug in the future. Thanks again for catching this, will release momentarily.